### PR TITLE
add Kronecker product of matrices

### PR DESCRIPTION
**Pull request by filiatra**

### CODE REVIEW

File: src/gsMatrix/gsSparseMatrix.h [hunk 1]
```diff
@@ -433,6 +433,40 @@ class gsSparseMatrix : public Eigen::SparseMatrix<T,_Options,_Index>
         return result;
     }
 
+    index_t bandWidth(index_t upto = std::numeric_limits<index_t>::max())  const
```
Comment (stefantakacs) [inline src/gsMatrix/gsSparseMatrix.h]: bandwidth is something else:
https://en.wikipedia.org/wiki/Band_matrix

Better: maxNonzerosPerCol

Fixing per row or per col is either inefficient for RowMajor or ColMajor

File: src/gsMatrix/gsSparseMatrix.h [hunk 2]
```diff
@@ -433,6 +433,40 @@ class gsSparseMatrix : public Eigen::SparseMatrix<T,_Options,_Index>
...
+        return nz;
+    }
+
+    /// Returns the Kronecker product of \a this with \a other
+    gsSparseMatrix kron(const gsSparseMatrix& other)  const
+    {
+        const index_t r  = this->rows(), c = this->cols();
+        const index_t ro = other.rows(), co = other.cols();
+        gsSparseMatrix result(r*ro, c*co);
+        result.reservePerColumn(this->bandWidth()*other.bandWidth());
```
Comment (stefantakacs) [inline src/gsMatrix/gsSparseMatrix.h]: reservePerColumn is either inefficient for RowMajor or for ColMajor.

File: src/gsTensor/gsTensorTools.h [hunk 3]
```diff
@@ -219,86 +219,4 @@ void flipTensorVector(const int dir,
...
-
-    \param[in]  m1     
-    \param[in]  m2
-    \param[out] result
-    \param[in] nzPerCol
-
-    \ingroup Tensor
-*/
-template<class T>
-void gsSparseKroneckerProduct(const gsSparseMatrix<T> & m1, const gsSparseMatrix<T> & m2,
```
Comment (stefantakacs) [inline src/gsTensor/gsTensorTools.h]: Is this unused?

**Comment by stefantakacs** [comment]
does not wor for rowMajor matrices.

File: src/gsTensor/gsTensorTools.h (hunk 3 above)
Comment (filiatra) [reply src/gsTensor/gsTensorTools.h]: yes

File: src/gsMatrix/gsSparseMatrix.h (hunk 1 above)
Comment (filiatra) [reply src/gsMatrix/gsSparseMatrix.h]: fixed, the allocation is accordingly..

**Comment by stefantakacs** [comment]
one dimension being 0 happened from time to time in devel.

**Pull request merged**